tox.ini

```diff
@@ -44,7 +44,8 @@ commands =
     # Clean up from previous tests
     python util.py clean-tests
     # Run tests
-    pytest -vv -rs --cov=sqlfluff --cov-report=xml {posargs: test}
+    pytest -vv -rs --cov=sqlfluff --cov-report=xml -m "not dbt_connection_failure" {posargs: test}
+    dbt{017,018,019,020,021,100}: pytest -vv -rs --cov=sqlfluff --cov-report=xml -m "dbt_connection_failure" {posargs: test} # Run the connection failure test in a separate invocation to reset the database connection
```

Comment:
This step has to run in separate invocation because dbt caches the connection to the database, and doesn't load the deliberately broken profile.

Reply:
There must be a way to reload the database connection in the same invocation, but I can't work out how!

Reply:
@NiallRees I actually have a slightly different suggestion here.

All this test really needs to do is validate how sqlfluff handles the exception when a connection error is returned by dbt. Rather than trying to use a different connection and having to split this out, we can just mock/monkey patch the return of the relevant function in the code to raise the connection error in this test. That way we can test how sqlfluff responds without having to actually trigger a real connection error 😄

[pytest-mock](https://pypi.org/project/pytest-mock/) is the package you want if you're interested.
I use it for things like testing how my code handles unlikely connection errors to databases/APIs 👍

Reply:
You can also use the built-in Python module `unittest.mock`. It provides a decorator, so it's pretty easy to use. https://otrabalhador.github.io/python-testing-by-examples/docs/en/mocking/essentials/patch.html

Reply:
As we have a database connection I thought we might as well break it for real, but perhaps mocking is the best option.

Reply:
I'm trying to run this test from the PR branch, and I'm getting an error about a missing file:
```
>       os.rename(src_fpath, target_fpath)
E       FileNotFoundError: [Errno 2] No such file or directory: 'plugins/sqlfluff-templater-dbt/test/fixtures/dbt/error_models/exception_connect_database.sql' -> '/Users/bhart/dev/sqlfluff/plugins/sqlfluff-templater-dbt/test/fixtures/dbt/dbt_project/models/my_new_project/exception_connect_database.sql'

plugins/sqlfluff-templater-dbt/test/templater_test.py:339: FileNotFoundError

 plugins/sqlfluff-templater-dbt/test/templater_test.py ⨯                                                                                                                                 100% ██████████
======================================================================================= short test summary info ========================================================================================
FAILED plugins/sqlfluff-templater-dbt/test/templater_test.py::test__templater_dbt_handle_database_connection_failure - FileNotFoundError: [Errno 2] No such file or directory: 'plugins/sqlfluff-temp...
```

@NiallRees: Do you have a file `plugins/sqlfluff-templater-dbt/test/fixtures/dbt/dbt_project/models/my_new_project/exception_connect_database.sql` that needs committing?

Reply:
Disregard my question about a missing file. My local clone was messed up somehow.

Reply:
We can now combine the two lines above, I think? (Now that the snow connection mock failure is in place.)